### PR TITLE
UCT/IB/MLX5: Move QP to RESET (instead of error) before cleaning the SRQ

### LINKED ISSUE #1963: Assertion `srq->tail == srq_info.tail' in */test_ucp_ep_force_disconnect.test/2
http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5001/label=hpc-test-node,worker=2/console
```
13:22:03 [ RUN      ] shm/test_ucp_ep_force_disconnect.test/2
13:22:03 [hpc-test-node:16818:0] ib_mlx5_log.c:109  FATAL: Error on QP 0x24dfb wqe[467]: Remote side aborted (synd 0x22 vend 0x99) opcode RECV
13:22:03 
13:22:03 /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/mlx5/ib_mlx5_log.c: [ uct_ib_mlx5_completion_with_err() ]
13:22:03       ...
13:22:03       104     }
13:22:03       105 
13:22:03       106     ucs_log(log_level, "Error on QP 0x%x wqe[%d]: %s (synd 0x%x vend 0x%x) opcode %s",
13:22:03 ==>   107             qp_num, wqe_counter, info, ecqe->syndrome, ecqe->vendor_err_synd,
13:22:03       108             uct_ib_mlx5_cqe_err_opcode(ecqe));
13:22:03       109 }
13:22:03       110 
13:22:03 
13:22:04 ==== backtrace ====
13:22:04  0 0x000000000002c0ca uct_ib_mlx5_completion_with_err()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/mlx5/ib_mlx5_log.c:107
13:22:04  1 0x000000000002cfc0 uct_ib_mlx5_check_completion()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/mlx5/ib_mlx5.c:258
13:22:04  2 0x0000000000048b44 uct_ib_mlx5_poll_cq()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/mlx5/ib_mlx5.inl:33
13:22:04  3 0x000000000003c7f4 uct_rc_mlx5_ep_t_cleanup()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/rc/accel/rc_mlx5_ep.c:461
13:22:04  4 0x00000000000571d6 ucs_class_call_cleanup_chain()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/ucs/type/class.c:50
13:22:04  5 0x0000000000046f80 uct_rc_mlx5_ep_t_delete()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/rc/accel/rc_mlx5_ep.c:483
13:22:04  6 0x0000000000015b81 ucp_ep_destroy_internal()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/ucp/core/ucp_ep.c:348
13:22:04  7 0x0000000000020999 ucp_worker_destroy_eps()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/ucp/core/ucp_worker.c:1132
13:22:04  8 0x0000000000020999 ucp_worker_destroy()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/ucp/core/ucp_worker.c:1140
13:22:04  9 0x000000000062c421 ucs::handle<ucp_worker*, void*>::release()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:356
13:22:04 10 0x000000000062c421 ucs::handle<ucp_worker*, void*>::reset()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:295
13:22:04 11 0x000000000062c421 ~handle()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:290
13:22:04 12 0x000000000062c421 ~pair()  /usr/include/c++/4.8.2/bits/stl_pair.h:96
13:22:04 13 0x000000000062c421 _Destroy<std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > >()  /usr/include/c++/4.8.2/bits/stl_construct.h:93
13:22:04 14 0x000000000062c421 __destroy<std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > >*>()  /usr/include/c++/4.8.2/bits/stl_construct.h:103
13:22:04 15 0x000000000062c421 _Destroy<std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > >*>()  /usr/include/c++/4.8.2/bits/stl_construct.h:126
13:22:04 16 0x000000000062c421 _Destroy<std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > >*, std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > >()  /usr/include/c++/4.8.2/bits/stl_construct.h:151
13:22:04 17 0x000000000062c421 std::vector<std::pair<ucs::handle<ucp_worker*, void*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*>, std::allocator<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > >, std::allocator<std::pair<ucs::handle<ucp_worker*, void*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*>, std::allocator<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > > > >::_M_erase_at_end()  /usr/include/c++/4.8.2/bits/stl_vector.h:1352
13:22:04 18 0x000000000062c421 std::vector<std::pair<ucs::handle<ucp_worker*, void*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*>, std::allocator<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > >, std::allocator<std::pair<ucs::handle<ucp_worker*, void*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*>, std::allocator<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > > > >::clear()  /usr/include/c++/4.8.2/bits/stl_vector.h:1126
13:22:04 19 0x000000000062c421 ucp_test_base::entity::cleanup()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/ucp/ucp_test.cc:463
13:22:04 20 0x000000000062c421 ucp_test_base::entity::~entity()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/ucp/ucp_test.cc:338
13:22:04 21 0x000000000057177e ucs::ptr_vector_base<ucp_test_base::entity>::release()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:223
13:22:04 22 0x000000000057177e ucs::ptr_vector<ucp_test_base::entity>::remove()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:246
13:22:04 23 0x0000000000565379 test_ucp_peer_failure::fail_receiver()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/ucp/test_ucp_peer_failure.cc:97
13:22:04 24 0x0000000000565379 ucs::entities_storage<ucp_test_base::entity>::sender()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test.h:134
13:22:04 25 0x0000000000565379 test_ucp_peer_failure::test_force_close()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/ucp/test_ucp_peer_failure.cc:197
13:22:04 26 0x000000000046e6c6 ucs::test_base::run()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test.cc:251
13:22:04 27 0x000000000046e6c6 ucs::test_base::TestBodyProxy()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test.cc:277
13:22:04 28 0x00000000004663c3 HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:3562
13:22:04 29 0x000000000045a83d testing::Test::Run()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:3635
13:22:04 30 0x000000000045a90c testing::TestInfo::Run()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:3812
13:22:04 31 0x000000000045aa6f testing::TestCase::Run()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:3930
13:22:04 32 0x000000000045f407 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:5802
13:22:04 33 0x000000000045f70b testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:5719
13:22:04 34 0x000000000040f633 RUN_ALL_TESTS()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest.h:20059
13:22:04 35 0x0000000000021b15 __libc_start_main()  ???:0
13:22:04 36 0x0000000000444fc9 _start()  ???:0
13:22:04 ===================
13:22:04 [1509535323.154023] [hpc-test-node:16818:0]    ucp_context.c:392  UCX  INFO  No accelerator devices are available
13:22:04 [1509535323.196428] [hpc-test-node:16818:0]    ucp_context.c:392  UCX  INFO  No accelerator devices are available
13:22:04 [1509535323.277014] [hpc-test-node:16818:0]    ucp_context.c:392  UCX  INFO  No accelerator devices are available
13:22:04 [1509535323.321537] [hpc-test-node:16818:0]    ucp_context.c:392  UCX  INFO  No accelerator devices are available
13:22:04 Sending notification to devendar@mellanox.com
13:22:07 [hpc-test-node:16818:0] Process frozen...
```

### PR DESCRIPTION
**Pull request by yosefe**
Fixes #1963

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6549/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@brminich pls take a look

File: src/uct/ib/rc/base/rc_ep.c [hunk 1]
```diff
@@ -291,7 +291,8 @@ ucs_status_t uct_rc_modify_qp(uct_rc_txqp_t *txqp, enum ibv_qp_state state)
     memset(&qp_attr, 0, sizeof(qp_attr));
     qp_attr.qp_state = state;
     if (ibv_modify_qp(txqp->qp, &qp_attr, IBV_QP_STATE)) {
-        ucs_warn("modify qp 0x%x to RESET failed: %m", txqp->qp->qp_num);
+        ucs_warn("modify qp 0x%x to state %d failed: %m", state,
```
Comment (brminich) [inline src/uct/ib/rc/base/rc_ep.c]: need to change parameters order: qp goes first

File: src/uct/ib/rc/accel/rc_mlx5_ep.c [hunk 2]
```diff
@@ -571,18 +574,40 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_ep_t, uct_iface_h tl_iface)
     return UCS_OK;
 }
 
+static void uct_rc_mlx5_ep_reset_qp(uct_rc_mlx5_ep_t *ep)
```
Comment (brminich) [inline src/uct/ib/rc/accel/rc_mlx5_ep.c]: do we need to call it for TM qp as well?

Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_ep.c]: yes, currently there is not SRQ cleanup flow for TM QP at all..

File: src/uct/ib/rc/base/rc_ep.c (hunk 1 above)
Comment (yosefe) [reply src/uct/ib/rc/base/rc_ep.c]: right

File: src/uct/ib/rc/accel/rc_mlx5_ep.c (hunk 2 above)
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_ep.c]: since it requires major change i prefer to do it in separate PR, and merge this one to make tests pass

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6551/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6556/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6559/ for details (Mellanox internal link).

**Pull request merged**